niceman/resource/singularity.py

```diff
@@ -0,0 +1,240 @@
+
+        # Stop the container.
+        self._runner.run(['singularity', 'instance.stop', self.name])
+
+        # Update status
+        self.id = None
+        self.status = None
+
+        # Remove temporary image file
+        os.remove(self.image_file)
+
+    def start(self):
+        """
+        Start a stopped container.
+        """
+        lgr.info("Singularity does not provide a start feature.")
+
+    def stop(self):
+        """
+        Stop a running container.
+        """
+        lgr.info("Singularity does not provide a stop feature.")
+
+    @borrowdoc(Resource)
+    def get_session(self, pty=False, shared=None):
+        if not self.get_instance_info():
+            self.connect()
+
+        if pty and shared is not None and not shared:
+            lgr.warning("Cannot do non-shared pty session for Singularity yet")
+        return (PTYSingularitySession if pty else SingularitySession)(
+            name=self.name
+        )
+
+    def get_instance_info(self):
+        """
+        Return basic information about a running container
+
+        Returns
+        -------
+        dict : instance info
+        """
+        try:
+            stdout, _ = self._runner.run(['singularity', 'instance.list'])
+        except CommandError:
+            return None
+
+        # Parse stdout to find the running instance.
+        # The output of instance.list is a table with columns:
+        #   DAEMON NAME, PID, and CONTAINER IMAGE
+        # Daemon names are unique on each server.
+        rows = stdout.split('\n')
+        for row in rows:
+            items = row.split()
+            if len(items) != 3:
+                # Skip header and trailing blank lines.
+                continue
+            if self.name == items[0] or self.id == items[0]:
+                return {
+                    'name': items[0],  # daemon name
+                    'pid': items[1],   # daemon process id
+                    'image': items[2]  # container image file
+                }
+
+
+@attr.s
+class SingularitySession(POSIXSession):
+    name = attr.ib()
+    _runner = Runner()
+
+    @borrowdoc(Session)
+    def _execute_command(self, command, env=None, cwd=None):
+        if env:
+            raise NotImplementedError("passing env variables to session execution")
+        if cwd:
+            raise NotImplementedError("handle cwd for singularity")
+        lgr.debug('Running command %r', command)
+        # If command is a string, convert it to a list
+        if isinstance(command, six.string_types):
+            command = command.split()
+        stdout, stderr = self._runner.run(['singularity', 'exec',
+            'instance://' + self.name] + command)
+
+        return (stdout, stderr)
+
+    @borrowdoc(Session)
+    def put(self, src_path, dest_path, uid=-1, gid=-1):
+        dest_dir, _ = os.path.split(dest_path)
+        self.mkdir(dest_dir, parents=True)
+        cmd = 'cat {} | singularity exec instance://{} tee {} > /dev/null'
```

Comment:
Isn't `cat IN | tee OUT >/dev/null` the same as `cat IN | cat >OUT`?

Reply:
It is. Will change

Reply:
eh, yeah, you'd have to do something ugly like `/bin/sh -c 'cat ... >OUT'`.  Sorry for the noise.

Reply:
No worries at all! As you can see from my having to re-edit my response that I had forgotten why I had done the tee thing in the first place.  :)